File: web/src/pose.ts
import { findHomography, type Point } from "./homography";
import { WHITE_COUNT } from "./keys";

export const WHITE_KEY_MM = 23.5;
// visible key depth of the instrument in front of the camera, measured from its labelled
// frames: the pose residual over 16 snaps bottoms out at a span-to-depth ratio of 7.2
export const KEYBED_DEPTH_MM = 118;
export const WHITE_KEY_COUNT = WHITE_COUNT;
export const DEPTH_UNITS = KEYBED_DEPTH_MM / WHITE_KEY_MM;
const SCAN_SAMPLES = 200;
const GOLDEN_ITERATIONS = 100;
const GOLDEN_RATIO = (Math.sqrt(5) - 1) / 2;

export interface PlanePose {
  focal: number;
  rotation: number[][];
  translation: number[];
  worldWidthMm: number;
  residual: number;
}

const WORLD_CORNERS: Point[] = [
  { x: 0, y: 0 },
  { x: WHITE_KEY_COUNT, y: 0 },
  { x: WHITE_KEY_COUNT, y: DEPTH_UNITS },
  { x: 0, y: DEPTH_UNITS },
];

function edgeLength(quad: Point[], index: number): number {
  return Math.hypot(quad[index].x - quad[0].x, quad[index].y - quad[0].y);
}

function roll(quad: Point[], by: number): Point[] {
  return quad.map((_, i) => quad[(i + by) % quad.length]);
}

// WORLD_CORNERS puts the 52 white keys on edge 0->1, so an image quad has to agree before it solves.
// The remaining 180 degree choice is the dragger's: handle 1 to 2 runs along the black keys.
export function canonicalQuad(quad: Point[]): Point[] {
  return edgeLength(quad, 1) >= edgeLength(quad, 3) ? quad : roll(quad, 1);
}

function inverseCalibrationHomography(
  focal: number,
  width: number,
  height: number,
  h: number[],
): number[][] {
  const cx = width / 2;
  const cy = height / 2;
  const b: number[][] = [[], [], []];
  for (let col = 0; col < 3; col += 1) {
    b[0][col] = (h[col] - cx * h[col + 6]) / focal;
    b[1][col] = (h[col + 3] - cy * h[col + 6]) / focal;
    b[2][col] = h[col + 6];
  }
  return b;
}

function column(m: number[][], j: number): number[] {
  return [m[0][j], m[1][j], m[2][j]];
}

function norm(v: number[]): number {
  return Math.hypot(v[0], v[1], v[2]);
}

function unit(v: number[]): number[] {
  const n = norm(v);
  return [v[0] / n, v[1] / n, v[2] / n];
}

function cross(a: number[], b: number[]): number[] {
  return [
    a[1] * b[2] - a[2] * b[1],
    a[2] * b[0] - a[0] * b[2],
    a[0] * b[1] - a[1] * b[0],
  ];
}

function orthogonalityResidual(r1: number[], r2: number[]): number {
  const dot = r1[0] * r2[0] + r1[1] * r2[1] + r1[2] * r2[2];
  return Math.abs(dot) + Math.abs(1 - norm(cross(r1, r2)));
}

// a rotation's columns are orthogonal AND the same length. The length test is what refuses a
// frontal view of the wrong aspect, where orthogonality alone is satisfied by any rectangle.
function poseResidual(b: number[][]): number {
  const c0 = column(b, 0);
  const c1 = column(b, 1);
  const n0 = norm(c0);
  const n1 = norm(c1);
  return (
    orthogonalityResidual(unit(c0), unit(c1)) +
    Math.abs(1 - Math.min(n0, n1) / Math.max(n0, n1))
  );
}

function goldenSection(
  fn: (f: number) => number,
  lo: number,
  hi: number,
): number {
  let a = lo;
  let b = hi;
  let c = b - GOLDEN_RATIO * (b - a);
  let d = a + GOLDEN_RATIO * (b - a);
  let fc = fn(c);
  let fd = fn(d);
  for (let i = 0; i < GOLDEN_ITERATIONS; i += 1) {
    if (fc < fd) {
      b = d;
      d = c;
      fd = fc;
      c = b - GOLDEN_RATIO * (b - a);
      fc = fn(c);
    } else {
      a = c;
      c = d;
      fc = fd;
      d = a + GOLDEN_RATIO * (b - a);
      fd = fn(d);
    }
  }
  return (a + b) / 2;
}

export function estimateFocal(
  imageCorners: Point[],
  width: number,
  height: number,
): number {
  const h = findHomography(WORLD_CORNERS, canonicalQuad(imageCorners));
  const lo = 0.3 * width;
  const hi = 3 * width;
  const residual = (f: number): number => {
    return poseResidual(inverseCalibrationHomography(f, width, height, h));
  };
  const step = (hi - lo) / SCAN_SAMPLES;
  let bestFocal = lo;
  let bestResidual = residual(lo);
  for (let i = 1; i <= SCAN_SAMPLES; i += 1) {
    const f = lo + i * step;
    const value = residual(f);
    if (value < bestResidual) {
      bestFocal = f;
      bestResidual = value;
    }
  }
  const goldenFocal = goldenSection(residual, lo, hi);
  const goldenResidual = residual(goldenFocal);
  return goldenResidual < bestResidual ? goldenFocal : bestFocal;
}

export function solvePose(
  imageCorners: Point[],
  width: number,
  height: number,
): PlanePose {
  const h = findHomography(WORLD_CORNERS, canonicalQuad(imageCorners));
  const focal = estimateFocal(imageCorners, width, height);
  const b = inverseCalibrationHomography(focal, width, height, h);
  const scale = norm(column(b, 0));
  const r1 = [b[0][0] / scale, b[1][0] / scale, b[2][0] / scale];
  const r2 = [b[0][1] / scale, b[1][1] / scale, b[2][1] / scale];
  const r3 = cross(r1, r2);
  const metricScale = WHITE_KEY_MM / scale;
  return {
    focal,
    rotation: [
      [r1[0], r2[0], r3[0]],
      [r1[1], r2[1], r3[1]],
      [r1[2], r2[2], r3[2]],
    ],
    translation: [
      b[0][2] * metricScale,
      b[1][2] * metricScale,
      b[2][2] * metricScale,
    ],
    worldWidthMm: WHITE_KEY_COUNT * WHITE_KEY_MM,
    residual: poseResidual(b),
  };
}

export function projectPoint(
  pose: PlanePose,
  u: number,
  v: number,
  width: number,
  height: number,
): Point {
  const xmm = u * WHITE_KEY_MM;
  const ymm = v * WHITE_KEY_MM;
  const r = pose.rotation;
  const t = pose.translation;
  const zc = r[2][0] * xmm + r[2][1] * ymm + t[2];
  return {
    x: width / 2 + (pose.focal * (r[0][0] * xmm + r[0][1] * ymm + t[0])) / zc,
    y: height / 2 + (pose.focal * (r[1][0] * xmm + r[1][1] * ymm + t[1])) / zc,
  };
}

File: tools/src/kvt/pose.py
"""Can this quad be a perspective view of the keybed rectangle? Port of web/src/pose.ts.

The keybed is 52 white keys of 23.5 mm by 150 mm deep. With a pinhole camera of unknown focal
length, the rectangle-to-image homography decomposes into two rotation columns that must be
orthonormal; the focal that makes them so is the camera's, and how far they stay from
orthonormal is the residual. Measured on the labelled frames it is under 0.44 for a real
keybed quad and over 18 for a broken one.
"""

from collections.abc import Callable
from dataclasses import dataclass

import cv2
import numpy as np

from kvt.dataset import canonical_quad

WHITE_KEY_MM = 23.5
# the instrument in front of the camera: a 61-key board, 36 white keys, depth measured from
# labelled frames at a span-to-depth ratio of 7.2; synthetic renders use their own 88-key geometry
KEYBED_DEPTH_MM = 118.0
WHITE_KEY_COUNT = 36
RENDER_WHITE_KEY_COUNT = 52
RENDER_DEPTH_UNITS = 150.0 / WHITE_KEY_MM
DEPTH_UNITS = KEYBED_DEPTH_MM / WHITE_KEY_MM
_WORLD = np.array(
    [[0.0, 0.0], [WHITE_KEY_COUNT, 0.0], [WHITE_KEY_COUNT, DEPTH_UNITS], [0.0, DEPTH_UNITS]],
    dtype=np.float32,
)
_SCAN_SAMPLES = 200
_GOLDEN_ITERATIONS = 100
_GOLDEN = (np.sqrt(5.0) - 1.0) / 2.0


@dataclass(frozen=True)
class PoseFit:
    focal: float
    residual: float


def _residual_at(h: np.ndarray, focal: float, cx: float, cy: float) -> float:
    b = np.empty((3, 3))
    b[0] = (h[0] - cx * h[2]) / focal
    b[1] = (h[1] - cy * h[2]) / focal
    b[2] = h[2]
    n1, n2 = float(np.linalg.norm(b[:, 0])), float(np.linalg.norm(b[:, 1]))
    r1, r2 = b[:, 0] / n1, b[:, 1] / n2
    # a rotation's columns are orthogonal AND the same length; the length test is what stops a
    # frontal view of the wrong aspect passing, where orthogonality alone says nothing
    return float(
        abs(np.dot(r1, r2))
        + abs(1.0 - np.linalg.norm(np.cross(r1, r2)))
        + abs(1.0 - min(n1, n2) / max(n1, n2))
    )


def _golden(fn: Callable[[float], float], lo: float, hi: float) -> float:
    a, b = lo, hi
    c = b - _GOLDEN * (b - a)
    d = a + _GOLDEN * (b - a)
    fc, fd = fn(c), fn(d)
    for _ in range(_GOLDEN_ITERATIONS):
        if fc < fd:
            b, d, fd = d, c, fc
            c = b - _GOLDEN * (b - a)
            fc = fn(c)
        else:
            a, c, fc = c, d, fd
            d = a + _GOLDEN * (b - a)
            fd = fn(d)
    return (a + b) / 2.0


def fit_pose(quad_px: np.ndarray, width: int, height: int) -> PoseFit:
    """Focal length and orthonormality residual for the quad read as the keybed rectangle."""
    ordered = canonical_quad(quad_px).astype(np.float32)
    h = cv2.getPerspectiveTransform(_WORLD, ordered)
    cx, cy = width / 2.0, height / 2.0

    def residual(focal: float) -> float:
        return _residual_at(h, focal, cx, cy)

    lo, hi = 0.3 * width, 3.0 * width
    step = (hi - lo) / _SCAN_SAMPLES
    best_focal, best = lo, residual(lo)
    for i in range(1, _SCAN_SAMPLES + 1):
        f = lo + i * step
        value = residual(f)
        if value < best:
            best_focal, best = f, value
    golden = _golden(residual, lo, hi)
    golden_value = residual(golden)
    if golden_value < best:
        best_focal, best = golden, golden_value
    return PoseFit(focal=float(best_focal), residual=float(best))
